packages/dds/tree/src/feature-libraries/modular-schema/schemaChangeFamily.ts

```diff
@@ -0,0 +1,278 @@
+            if (change.schema !== undefined) {
+                schemaId = change.schema;
+            }
+
+            for (const key of Object.keys(change.fields)) {
+                fieldChanges.add(brand(key), change.fields[key]);
+            }
+        }
+
+        assert(schemaId !== undefined, "Need schema ID for root");
+        const schema = this.getTreeSchema(schemaId);
+
+        const composedFields: FieldChangeMap = {};
+        for (const field of fieldChanges.keys()) {
+            const childSchema = this.getChildSchema(schema, field);
+            const composedField = this.getChangeHandler(schema, field).rebaser.compose(
+                fieldChanges.get(field),
+                (...childChanges) => this.composeI(childSchema, childChanges),
+            );
+
+            // TODO: Could optimize by checking that composedField is non-empty
+            composedFields[field as string] = composedField;
+        }
+
+        const result: NodeChangeset = { fields: composedFields };
+        if (staticSchemaId === undefined) {
+            result.schema = schemaId;
+        }
+        return result;
+    }
+
+    invert(changes: NodeChangeset): NodeChangeset {
+        return this.invertI(this.rootSchema, changes);
+    }
+
+    private invertI(staticSchemaId: TreeSchemaIdentifier | undefined, changes: NodeChangeset): NodeChangeset {
+        const schemaId = staticSchemaId ?? changes.schema;
+        assert(schemaId !== undefined, "Need schema ID for root");
+        const schema = this.getTreeSchema(schemaId);
+
+        const fields: FieldChangeMap = {};
+
+        for (const field of Object.keys(changes.fields)) {
+            const childSchema = this.getChildSchema(schema, brand(field));
+            fields[field] = this.getChangeHandler(schema, brand(field)).rebaser.invert(
+                changes.fields[field],
+                (nodeChanges) => this.invertI(childSchema, nodeChanges));
+        }
+
+        const result: NodeChangeset = { fields };
+        if (staticSchemaId === undefined) {
+            result.schema = schemaId;
+        }
+        return result;
+    }
+
+    rebase(change: NodeChangeset, over: NodeChangeset): NodeChangeset {
+        return this.rebaseI(this.rootSchema, change, over);
+    }
+
+    private rebaseI(
+        staticSchemaId: TreeSchemaIdentifier | undefined,
+        change: NodeChangeset,
+        over: NodeChangeset,
+    ): NodeChangeset {
+        const schemaId = staticSchemaId ?? change.schema;
+        assert(schemaId !== undefined, "Need schema ID for root");
+        const schema = this.getTreeSchema(schemaId);
+
+        const fields: FieldChangeMap = {};
+
+        for (const field of Object.keys(change.fields)) {
+            const baseChanges = over.fields[field];
+            if (baseChanges === undefined) {
+                fields[field] = change.fields[field];
+            } else {
+                const childSchema = this.getChildSchema(schema, brand(field));
```

Comment:
All this schema use in here is problematic. I think we want this code to work in terms of stored schema, and at any given point in time there isn't a correct stored schema: thats branch specific.

We either need to pass in state information about the current schema (either explicitly, or implicitly by capturing them in the NodeChangeSet), or not depend on it in here (ex: if we put field kinds (or even full schema) on paths when building the edits, we shouldn't need them in here.